File: order/order-domain/src/main/java/shopping/order/domain/Product.java
package shopping.order.domain;

import java.math.BigInteger;

public class Product {

    private final long id;
    private final String name;
    private final BigInteger price;
    private final String imageUrl;

    public Product(long id, String name, BigInteger price, String imageUrl) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.imageUrl = imageUrl;
    }

    public ReceiptProduct purchase(int quantity) {
        return new ReceiptProduct(id, name, calculate(quantity), imageUrl, quantity);
    }

    public BigInteger calculate(int quantity) {
        return price.multiply(BigInteger.valueOf(quantity));
    }

    public long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public BigInteger getPrice() {
        return price;
    }

    public String getImageUrl() {
        return imageUrl;
    }

}

File: src/main/kotlin/shopping/domain/Product.kt
package shopping.domain

class Product(
    val id: Long,
    val name: ProductName,
    val price: Int,
    val imageUrl: String,
) {
    init {
        require(imageUrl.startsWith("http://") || imageUrl.startsWith("https://")) {
            "[ERROR] imageUrl must start with http:// or https://"
        }
    }
}
